libraries/chain/controller.cpp

```diff
@@ -1257,6 +1245,64 @@ struct controller_impl {
       }
    }
 
+   bool should_transition_to_savanna(const block_state_ptr&) {
+      return false;
+   }
+   bool should_transition_to_savanna(const block_state_legacy_ptr& bsp) {
+      return bsp->header.contains_header_extension(instant_finality_extension::extension_id());
+   }
+
+   void transition_to_savanna() {
+      assert(chain_head.header().contains_header_extension(instant_finality_extension::extension_id()));
+      // copy head branch branch from legacy forkdb legacy to savanna forkdb
+      fork_database_legacy_t::branch_t legacy_branch;
+      block_state_legacy_ptr legacy_root;
+      fork_db.apply_l<void>([&](const auto& forkdb) {
+         legacy_root = forkdb.root();
+         legacy_branch = forkdb.fetch_branch(forkdb.head()->id());
+      });
+
+      assert(!!legacy_root);
+      assert(read_mode == db_read_mode::IRREVERSIBLE || !legacy_branch.empty());
+      ilog("Transitioning to savanna, IF Genesis Block ${gb}, IF Critical Block ${cb}", ("gb", legacy_root->block_num())("cb", chain_head.block_num()));
+      auto new_root = block_state::create_if_genesis_block(*legacy_root);
+      fork_db.switch_from_legacy(new_root);
+      fork_db.apply_s<void>([&](auto& forkdb) {
+         block_state_ptr prev = forkdb.root();
+         for (auto bitr = legacy_branch.rbegin(); bitr != legacy_branch.rend(); ++bitr) {
+            const bool skip_validate_signee = true; // validated already
+            auto new_bsp = std::make_shared<block_state>(
+                  *prev,
+                  (*bitr)->block,
+                  protocol_features.get_protocol_feature_set(),
+                  validator_t{}, skip_validate_signee);
+            // legacy_branch is from head, all should be validated
+            assert((*bitr)->action_receipt_digests);
+            auto action_mroot = calculate_merkle(*(*bitr)->action_receipt_digests);
```

Comment:
I don't believe you'll be able to reuse the `action_receipt_digests` since they will be different for legacy and Savanna blocks.

Reply:
Is that coming in your PR? If so, should you update this as part of that work? Nothing I can do about that now, right?

Reply:
Well I think it means that we will need to recalculate the `action_receipt_digests` anyways. Do we have the traces stored somewhere in the block_state_legacy? Otherwise we need to compute both when applying the block during the transition. 
I was looking at your PR, and it is not clear what I can use to determine which type(s) of `action_receipt_digests` to compute?